pennylane/devices/qubit/preprocess.py

```diff
@@ -0,0 +1,220 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""This module contains functions for preprocessing `QuantumScript`s to ensure
+that they are supported for execution by a device."""
+
+import pennylane as qml
+
+from pennylane.operation import Observable, Tensor
+from pennylane.ops import Sum
+from pennylane.measurements import (
+    MidMeasureMP,
+    ExpectationMP,
+    ShadowExpvalMP,
+)
+from pennylane import DeviceError
+
+# Update observable list. Current list is same as supported observables for
+# default.qubit.
+_observables = {
+    "PauliX",
+    "PauliY",
+    "PauliZ",
+    "Hadamard",
+    "Hermitian",
+    "Identity",
+    "Projector",
+    "SparseHamiltonian",
+    "Hamiltonian",
+    "Sum",
+    "SProd",
+    "Prod",
+    "Exp",
+    "Evolution",
+}
+
+
+def _stopping_condition(op):
+    """Specify whether or not an Operator object is supported by the device"""
+    return getattr(op, "has_matrix", False) and qml.matrix(op).shape[-1] % 2 == 0
+
+
+def _supports_observable(observable):
+    """Checks if an observable is supported by this device.
+
+    Args:
+        observable (type or str): observable to be checked
+
+    Returns:
+        bool: ``True`` iff supplied observable is supported
+    """
+    if isinstance(observable, type) and issubclass(observable, Observable):
+        observable = observable.__name__
+    if isinstance(observable, str):
+        return observable in _observables
+
+    return False
+
+
+def expand_fn(circuit, max_expansion=10):
+    """Method for expanding or decomposing an input circuit.
+
+    This method expands the tape if:
+
+    - nested tapes are present,
```

Comment:
I wonder.. @albi3ro didn't we want to get rid of nested tapes? Afaik only the QuantumTape class easily supports this kind of behaviour and we want to phase it out. Probably discussion for another PR, but I wouldn't mind dropping this stuff. FWIW QNodes do not nested-queue like they used to anymore.

Reply:
I don't think nested tapes are a thing anymore. I'm actually not sure if they were ever really much of a thing.

Reply:
anyway - we don't explicitly check for this, and we might change this check altogether re: the discussion on checking the matrices

Reply:
Updated the doc :)